### DISCUSSION TITLE
Next.js. multer not upload on deploy, but upload on localhost

### DISCUSSION THREAD

**Discussion by Xazgl** [vercel/next.js#50014] [Help]
### Summary

Find the reason why I can't save files on the server (But localhost saves ) images to the /public/uploads folder. Uploads exists and its access rights are drwxrwxrwx. There are no errors, the api writes that everything is successful, it saves the file name to the database, but there are no files in the /public/uploads folder itself
Please, help me!

### Additional information

```js
https://stackoverflow.com/questions/76278691/why-multer-not-save-file-on-server-nginx-next-js
```

**Comment by icyJoseph** [comment]
How are you hosting and running the project? Is it on Vercel hosting?

**Reply by Xazgl** [reply]
MY dockerfile : 
```
FROM node:lts-alpine AS builder


# Create app directory
WORKDIR /app

# A wildcard is used to ensure both package.json AND package-lock.json are copied
COPY package*.json ./
COPY prisma ./prisma/


# Install app dependencies
RUN npm install
RUN npx prisma generate


COPY . .

RUN npm run build

FROM node:lts-alpine

WORKDIR /app

COPY --from=builder /app/node_modules ./node_modules
COPY --from=builder /app/package*.json ./
COPY --from=builder /app/.next ./.next
COPY --from=builder /app/public ./public
COPY --from=builder /app/prisma ./prisma


# Install Redis
RUN apk add --update redis


# # Stop Redis service
# RUN pkill redis-server && redis-cli shutdown

# Copy Redis configuration file
COPY redis.conf /usr/local/etc/redis/redis.conf  


# Set ownership and permissions for Redis configuration file
RUN chown redis:redis /usr/local/etc/redis/redis.conf

# Set Redis configuration
ENV REDIS_HOST=domen
ENV REDIS_PORT=6379

ENV NODE_ENV=production



EXPOSE 3000

CMD ["npx", "prisma", "migrate", "deploy", "--preview-feature"]

```

**Reply by Xazgl** [reply]
@icyJoseph, Please tell me, they told me (I inserted the quote below), it turns out I can't upload files to Next in any way with something like multer.js after deployment ? Even if the file will not be displayed until the full build of the project again. The idea is to upload it as easily as possible into an existing project, through the administration panel I made, and then the user can launch the "Site Assembly" and a picture for the promotion will appear 

  Was told this : 

> You won't be able to upload new files to a live Vercel Project outside of build and deployment time. But you can use an external service to store files uploaded via your site.
> 
> Related docs and guides:
> 
> [How to Upload and Store Files with Vercel](https://vercel.com/guides/how-to-upload-and-store-files-with-vercel)
> [How can I use files in Serverless Functions on Vercel?](https://vercel.com/guides/how-can-i-use-files-in-serverless-functions)
> [Vercel Blob](https://vercel.com/docs/storage/vercel-blob#) (currently in private beta)



MY API: 
 ```
import type { NextApiRequest, NextApiResponse } from 'next'
import db, { Offer } from '../../prisma';
import nextConnect from 'next-connect';
import multer from 'multer';
import checkSession from "../../src/services/checkCookie";
import helmet from "helmet";
import { z, ZodError } from 'zod';

const upload = multer({
    storage: multer.diskStorage({
        destination: './public/uploads',
        filename: (req, file, cb) => cb(null, file.originalname),
    }),
});

const apiRoute = nextConnect<NextApiRequest & { file?: Express.Multer.File }, NextApiResponse>({
    onError(error, req, res) {
        res.status(501).json({ error: `Sorry something Happened! ${error.message}` });
    },
    onNoMatch(req, res) {
        res.status(405).json({ error: `Method '${req.method}' Not Allowed` });
    },
});

apiRoute.use(helmet());
apiRoute.use(upload.single('image'));

apiRoute.post(async (req, res, next) => {
    const token = req.cookies['sid']
    const admin = await checkSession(token)
    if (admin) {
        return next()
    }
    next(new Error('Auth required'))
}, async (req, res) => {
    try {
        const img: string = req.file ? req.file.filename : ''
        const loginSchema = z.object({
            title: z.string().min(2).max(50),
            shortDesc: z.string().min(2).max(150),
            description: z.string(),
            filterMainPeople: z.string().min(2).max(60),
            detailFilterBrand: z.string().min(2).max(60),
            detailFilterMode: z.string().min(2).max(60),
            price: z.string().max(60),
        })
        const {
            title, shortDesc, description, filterMainPeople, detailFilterBrand, detailFilterMode, price
        } = loginSchema.parse(req.body)
        const newOffer = await db.offer.create({
            data: {
                title,
                shortDesc,
                description,
                filterMainPeople,
                detailFilterBrand,
                detailFilterMode,
                img,
                price,
                active: true,
            }
        })
        res.status(200).send(newOffer)
    } catch (error) {
        console.error(error)
        if (error instanceof ZodError) {
            return res.status(404).send({ message: "error validation" })
        }
        res.status(500).send({ message: "server error" })
    }
});
```

**Reply by icyJoseph** [reply]
Ah, you have to read carefully... if you deploy to Vercel, your app will be a serverless app, and those can't modify the File System, for various reasons.

> On a serverless deploy, each API route, and server side code, runs in a lambda, as a task, then it is killed, and with it, the memory it used, etc, https://github.com/orgs/vercel/discussions/241#discussioncomment-363002

If you are deploying this on your own environment, like your own machine let's say, then it'll work. For example:

https://github.com/vercel/next.js/assets/21013447/2ae580cf-9642-42b1-939d-e73311a44e5a

Notice that I am using `process.cwd()` in combination with `resolve` from `path`, to correctly find the public directory.

Again, if you are in control of the server deploy, and it is not some kind of serverless operation/task, it should work. However...

#  ❌ Although you can save them, you can't reach them

- https://nextjs.org/docs/app/building-your-application/optimizing/static-assets

> Only assets that are in the public directory at [build time](https://nextjs.org/docs/app/api-reference/next-cli#build) will be served by Next.js. Files added at runtime won't be available. We recommend using a third party service like [AWS S3](https://aws.amazon.com/s3/) for persistent file storage.

For the files to be available, you'll have to re-build the project, so it is better to use some other kind of storage. Not sure why you want to use the public folder, I thought I read you just wanted to upload them, and if so, the above works, but if you want to server them afterwards, it won't work.

https://github.com/vercel/next.js/assets/21013447/b3e80f81-d8d7-4d25-9b33-43832717e53d

Though I wonder, if you saved them  to a tmp folder and exposed them through an API route mmmm, like `api/tmp/[file-name]` and this API would read and serve the content's if found 🤔 pf course, would not work in serverless environments. **It could also be an attack vector and someone could trick the API route into reading secrets and stuff** 😱 better to store the files somewhere else!

**Reply by Xazgl** [reply]
If I understood correctly, then the first method will help save the file to a shared directory, but I won't be able to get it until the next build? If yes, then I can run a project rebuild, and for example bind a public/upload link to an image inside the database/newSale.webp and upload this file to this folder and after a new build, the link will stop being a bit and will see a new image in the folder?


 I could use Amazon, but in this project I would like to minimize the financial part, but as I understand it, all such services will be paid?

**Comment by Tommy1223454** [comment]
If the upload path behaves differently between localhost and deploy, I’d try to separate two concerns:

1. transport / parsing
2. trust + persistence

In production I usually avoid treating “file arrived successfully” as the same thing as “file is safe to keep”. A cleaner pattern is parse -> inspect -> then store/promote.

I built an OSS package for that inspection step in Node.js:
https://github.com/pompelmi/pompelmi

It’s useful when you want server-side checks before persisting untrusted uploads, instead of relying only on the fact that Multer accepted the request.
